### DISCUSSION TITLE
Dev container, setup instructions, and virtual environment: conda vs venv

### DISCUSSION THREAD

**Discussion by BurtHarris** [microsoft/promptflow#3590] [General]
Trying to get started in this repro.  The instructions in [*dev_setup.md*](https://github.com/microsoft/promptflow/blob/main/docs/dev/dev_setup.md) don't seem compatible with the base image used in the *devcontainer.json* file. 
The dev_setup instructions start with a **conda** command, but conda isn't installed in the image built by the dev container.

Not having much experience with python, nor with dev containers, this has been a confusing experience, even getting many of the examples to run has been painful.   I've tried using the dev container, both locally and in a github codespace.   

Should I use a standard python venv virtual environment?  Or install conda?   Or is a virtual environment needed at all, since I'm working in a dev container?

To further complicate things, building a venv from the command line inside vscode doesn't seem to let vscode know the environment should be activated, but if I used the vscode's tool to create the venv, that seems a step forward.

---
```
root@fd187e492970:/workspaces/promptflow# conda pip install -r requirements.txt
bash: conda: command not found
root@fd187e492970:/workspaces/promptflow# pip install -r requirements.txt
ERROR: Could not open requirements file: [Errno 2] No such file or directory: 'requirements.txt'

[notice] A new release of pip is available: 23.0.1 -> 24.1.2
[notice] To update, run: pip install --upgrade pip
root@fd187e492970:/workspaces/promptflow# pip install --upgrade pip
Requirement already satisfied: pip in /usr/local/lib/python3.9/site-packages (23.0.1)
Collecting pip
  Downloading pip-24.1.2-py3-none-any.whl (1.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 5.7 MB/s eta 0:00:00
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.0.1
    Uninstalling pip-23.0.1:
      Successfully uninstalled pip-23.0.1
Successfully installed pip-24.1.2
WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
root@fd187e492970:/workspaces/promptflow# conda create -n pf-examples python=3.9
bash: conda: command not found
root@fd187e492970:/workspaces/promptflow# 
 *  History restored 

root@fd187e492970:/workspaces/promptflow#  
```

**Comment by 0mza987** [comment]
Hi @BurtHarris 

Thanks for the feedback. 

- `dev_setup.md` is a guide to set up the development environment for promptflow project itself, it's only needed when you want to contribute features to prompt flow. This is for developers.
- Dev container is a convenient tool to set up prompt flow to develop your own LLM application that's based on prompt flow. This is for users. You can refer to [Quick start — Prompt flow documentation](https://microsoft.github.io/promptflow/how-to-guides/quick-start.html) to check how to use prompt flow. Dev container basically equals to `installation` part. You can skip installation if you are already in a dev container.

**Comment by BurtHarris** [comment]
Thank you.  Am I to understand that the promptflow dev team does not use dev containers?   They use conda environments instead?  

I'll certainly go back and retry with the quick start instructions thank.    It is my hope to contribute, starting with documentation improvements, but not the core code (I've got a lot to learn about python.)  

I run vscode on win11.   My initial intent was simply to fork and look around. 

1) Using vscode I cloned my forked repo (to my dev drive E:) `> git clone (recursive)`
2) vscode offers to open the cloned repo. I accept.
3) vscode noticed the .devcontainer folder and offers to reopen the project in a container.  **I don't know much about dev containers, but I accept. **
4) I start a vscode terminal window:
> `pip install promptflow promptflow-tools`.  (OK, I'm surprised that setup isn't done in .devcontainer.json).   

Among the log, it says:   `WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour` [sic] `with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv`.  This message sent me looking for what command to use to create the virtual environment, and eventually leading me to the `conda` command in the developer setup document, but this time I will ignore it.

> `pf flow init --flow ./my_chatbot --type chat`

5)  At this point, working with a new tool (pf command), I want to see what changed.   My instinct is to open the vscode source control panel.    It displays a warning I've not seen before about an 'unsafe repo'

![image](https://github.com/user-attachments/assets/8d654304-e085-42cb-bcbc-a1c1f4320965)

Note:  The instructions in quick start's *Setup a connection for your API key* were a little unclear here, my apprentice didn't understand that the *OpenAI* and *Azure OpenAI* options are alternatives.   I spotted it, as I had read of the concept of *connection* in this context.  

I'll continue to move ahead with this, perhaps offer a PR with improvements to the quick start documentation.

**Comment by BurtHarris** [comment]
I guess having the .devcontainer setup at the root of the project had the process confused, perhaps this was because I git cloned from within vscode, rather than another way.

**Comment by 0mza987** [comment]
- Most of our dev team members use conda to set up dev environment, like descirbed in `dev_setup.md`.
- You can leave dev container alone if you are not familiar with it.
- `Quick start` documentation and the whole doc site can help you familiarize yourself with prompt flow project.
- You are welcome to contribute doc/code to prompt flow project, just remember to create an issue thread in [Issues · microsoft/promptflow](https://github.com/microsoft/promptflow/issues) and link PR there.
- Happy exploring!
